lib/cc/cli/version_checker.rb

```diff
@@ -0,0 +1,83 @@
+require "uuid"
+
+module CC
+  module CLI
+    class VersionChecker
+      include CC::CLI::Output
+
+      VERSION_CHECK_TIMEOUT = 60 * 60 # 1 Hour in seconds
+      DEFAULT_VERSIONS_URL = "https://versions.codeclimate.com".freeze
+
+      def check
+        return unless check_version?
+
+        print_new_version_message if outdated?
+      end
+
+      private
+
+      def check_version?
+        first_arg = ARGV.first
+        if %w[--check-version].include? first_arg
+          ARGV.shift
+          true
+        else
+          false
+        end
+      end
+
+      def outdated?
+        api_response["outdated"] == true
+      end
+
+      def latest_version
+        api_response["latest"]
+      end
+
+      def print_new_version_message
+        warn "A new version (v#{latest_version}) is available"
+      end
+
+      def api_response
+        @api_response ||=
+          begin
+            JSON.parse(api_response_body)
+          rescue JSON::ParserError => error
+            CLI.debug(error)
+            # We don't know so pretend all is peachy. We'll try again next time.
```

Comment:
`Ctrl-C` and similar wouldn't be handled by `rescue => ex` because they don't descend from `StandardError`, specifically for that reason.

Because there is a UX constraint that this object should be best-effort and can never cause the CLI itself to error, I believe you will need a `rescue => ex` in `#check`. If it makes you feel better to still rescue the parse error or http error internally, that's fine with me; I would just find it to be unnecessary clutter at that point.

(Sorry I didn't clarify the UX constraint in my original comment, I only just now realized that are hands are a bit tied in that way.)